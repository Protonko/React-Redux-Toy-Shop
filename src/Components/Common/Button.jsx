import React from 'react';

export const Button = ({title, onClick}) => (
    <button className="button" onClick={onClick}>
        {title}
    </button>
);
