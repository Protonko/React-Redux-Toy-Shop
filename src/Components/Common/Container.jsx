import React from 'react';

export const Container = ({ children, classNames }) => (
    <div className={classNames}>{children}</div>
);
