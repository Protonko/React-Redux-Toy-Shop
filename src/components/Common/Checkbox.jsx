import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export const Checkbox = ({text, classNamesText = [], onChange}) => {
  const classNamesCheckboxText = classNames(
    'checkbox__text',
    ...classNamesText,
  );

  return (
    <label className="checkbox">
      <input type="checkbox" className="checkbox__input" onChange={onChange} />
      <span className={classNamesCheckboxText}>{text}</span>
    </label>
  );
}

Checkbox.propTypes = {
  text: PropTypes.string,
  classNamesAdditional: PropTypes.array,
};
