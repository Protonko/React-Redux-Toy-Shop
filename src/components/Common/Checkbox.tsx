// types
import {TClassName} from 'models/types';

import React, {FC} from 'react';
import classNames from 'classnames';

interface ICheckboxProps {
  text?: string,
  classNamesText?: TClassName,
  checked?: boolean,
  onChange: () => void,
}

export const Checkbox: FC<ICheckboxProps> = ({
  text,
  classNamesText = [],
  checked,
  onChange
}) => {
  const classNamesCheckboxText = classNames(
    'checkbox__text',
    ...classNamesText,
  );

  return (
    <label className="checkbox">
      <input type="checkbox" className="checkbox__input" checked={checked} onChange={onChange} />
      <span className={classNamesCheckboxText}>{text}</span>
    </label>
  );
}
