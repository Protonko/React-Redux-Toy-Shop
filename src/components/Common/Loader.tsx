import React, {FC} from 'react';

export const Loader: FC = () => (
    <div className="loader">
        <div className="loader__circle" />
        <div />
    </div>
);
