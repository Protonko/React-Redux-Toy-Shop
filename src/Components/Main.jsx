import React from 'react';
import Filter from '../containers/Filter';
import {Container} from './Common/Container';
import {Content} from './Content';
import {Loader} from './Common/Loader';

export const Main = props => {
    return (
        <main className="page__content">
            <Container classNames="container page__content-wrapper">
                {!props.isLoaded
                    ? <Loader />
                    : (
                        <>
                            <Content {...props} />
                            <Filter />
                        </>
                    )
                }
            </Container>
        </main>
    );
}
