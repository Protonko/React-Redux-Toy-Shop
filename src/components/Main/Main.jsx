import React from 'react';
import PropTypes from 'prop-types';
import {Loader} from 'components/Common/Loader';
import {Notification} from 'components/Common/Notification';
import Filter from 'containers/Filter';
import Cards from 'containers/Cards';

export const Main = ({isLoaded, products, selectedProducts}) => {
  if (isLoaded) {
    return (
      <>
        {products.length ? (
          <div className="page__content-main">
            <Cards products={products} selectedProducts={selectedProducts} />
          </div>
        ) : <Notification text="Nothing found" />}
        <Filter />
      </>
    )
  } else {
    return <Loader />;
  }
}

Main.propTypes = {
  isLoaded: PropTypes.bool,
  products: PropTypes.array,
  selectedProducts: PropTypes.array,
};
