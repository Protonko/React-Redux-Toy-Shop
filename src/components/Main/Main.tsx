// types
import {IProduct} from 'models/interfaces';

import React, {FC, useCallback, useEffect} from 'react';
import {usePagination} from 'hooks/usePagination';
import {Loader} from 'components/Common/Loader';
import {Notification} from 'components/Common/Notification';
import Filter from 'containers/Filter';
import Cards from 'containers/Cards';

export interface IProps {
  isLoaded: boolean,
  products: Array<IProduct>,
  selectedProducts: Array<IProduct>,
  setProducts: (page?: number) => void,
  setSearchQuery: (value: string) => void,
  page: number,
}

export const Main: FC<IProps> = ({
  isLoaded,
  products,
  selectedProducts,
  setProducts,
  setSearchQuery,
  page,
}) => {
  const fetchProducts = useCallback(() => {
    setProducts();
  }, [setProducts]);

  const resetSearchQuery = useCallback(() => {
    setSearchQuery('');
  }, [setSearchQuery])

  const handleLoadMore = () => {
    setProducts(page + 1);
  }

  useEffect(() => {
    if (!products.length) {
      fetchProducts();
    }

    return () => {
      resetSearchQuery();
    }
  }, [fetchProducts, resetSearchQuery]);

  usePagination(handleLoadMore);

  if (isLoaded) {
    return (
      <>
        {products.length ? (
          <div className="page__content-main">
            <Cards products={products} selectedProducts={selectedProducts} />
          </div>
        ) : <Notification text="Nothing found" />}

        <Filter />
      </>
    )
  } else {
    return <Loader />;
  }
}
