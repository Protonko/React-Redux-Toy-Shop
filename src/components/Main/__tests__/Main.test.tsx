import React from 'react';
import {shallow, mount} from 'enzyme';
import {products, selectedProducts} from 'static/__test__/card';
import {Main, IProps} from 'components/Main/Main';
import {Loader} from 'components/Common/Loader';

const ELEMENT_SELECTORS = {
  container: 'page__content-main',
  loader: Loader,
}

const shallowMainComponent = (props: IProps) => (
  shallow(<Main {...props} />)
);

const mountMainComponent = (props: IProps) => (
  mount(<Main {...props} />)
);

describe('Main', () => {
  let props: IProps;
  let componentWillUnmount: jest.Mock<string, []>;

  beforeEach(() => {
    props = {
      isLoaded: true,
      products,
      selectedProducts,
      setProducts: () => {},
      setSearchQuery: () => {},
      page: 0,
      errorMessage: null,
    };
    componentWillUnmount = jest.fn(() => '');
  });

  it('Match snapshot', () => {
    const component = shallowMainComponent(props);

    expect(component).toMatchSnapshot();
  });

  it('Should render Main component', () => {
    const component = shallowMainComponent(props);
    const wrapper = component.find(`.${ELEMENT_SELECTORS.container}`);

    expect(wrapper.length).toBe(1);
  });

  it('Should render Loader if data did not load', () => {
    props = {...props, isLoaded: false};
    const component = mountMainComponent(props);
    const loader = component.find(ELEMENT_SELECTORS.loader);

    expect(loader.length).toBe(1);
  });

  it('Should reset search query on unmout', () => {
    expect(componentWillUnmount()).toBe('');
  });
});
