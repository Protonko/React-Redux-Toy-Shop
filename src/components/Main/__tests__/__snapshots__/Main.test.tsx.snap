// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Match snapshot 1`] = `
<Fragment>
  <div
    className="page__content-main"
  >
    <Connect(Component)
      products={
        Array [
          Object {
            "category": "category",
            "checked": true,
            "id": 0,
            "price": 10,
            "rating": 5,
            "sale": 0.9,
            "title": "title",
          },
          Object {
            "category": "category 1",
            "checked": true,
            "id": 1,
            "price": 1,
            "rating": 3,
            "sale": 0.5,
            "title": "title 1",
          },
        ]
      }
      selectedProducts={
        Array [
          Object {
            "category": "category",
            "checked": true,
            "id": 0,
            "price": 10,
            "rating": 5,
            "sale": 0.9,
            "title": "title",
          },
        ]
      }
    />
  </div>
  <Connect(Component) />
</Fragment>
`;
